### PR TITLE
Pre-lending protocol refactoring 2

### PR DESCRIPTION
**Pull request by ximinez**
## High Level Overview of Change

This PR is number two of four which contain several small changes that are used as the baseline of #5270. Since they are significant and useful, I created this separate PR to get them merged sooner.

The base of this branch is #5590. The expectation is that they'll be merged in order.

(This one is a pure refactor, and does not add or change functionality.)

### Context of Change

#5270 is large and complicated, and I had to lay some groundwork before I could implement the feature.

### Type of Change

- [X] Refactor (non-breaking change that only restructures code)

## Before / After

Highlights:
* Refactors `parseLeaf` to break the handlers for `STI_UINT16` and `STI_UINT32` into separate helper functions. This is mainly to improve readability and add some reusability. It was going to be the base for functionality that I abandoned as unnecessary.

### CODE REVIEW

**Comment by ximinez** [comment]
> This patch will add a new STParseJSON_test.cpp and should give you 100% on your changes. [gist.github.com/dangell7/a129e1032a88d6c13e64c57e42b2427a](https://gist.github.com/dangell7/a129e1032a88d6c13e64c57e42b2427a)

Thanks for the patch!

I verified that the code was moved over with no changes. I also renamed the file and class to `STParsedJSON_test` to match the existing class.

(In the future, it may save us both time if you push a commit to a new branch, and just give me the commit ID, which I can then cherry-pick.)

**Comment by dangell7** [comment]
> > This patch will add a new STParseJSON_test.cpp and should give you 100% on your changes. [gist.github.com/dangell7/a129e1032a88d6c13e64c57e42b2427a](https://gist.github.com/dangell7/a129e1032a88d6c13e64c57e42b2427a)
> 
> Thanks for the patch!
> 
> I verified that the code was moved over with no changes. I also renamed the file and class to `STParsedJSON_test` to match the existing class.
> 
> (In the future, it may save us both time if you push a commit to a new branch, and just give me the commit ID, which I can then cherry-pick.)

Is `if (value.isString())` inside of `parseUnsigned` a possible code path? It looks like its not but maybe we want to keep it for defensive purposes?

If its not testable and you want to keep it can we add an exclude?

```
// LCOV_EXCL_START
if (value.isString())
...
// LCOV_EXCL_STOP
```

**Review by Bronek — APPROVED** [review]
Minor comment aside (which I do not feel strongly about, but some response would be appreciated) this is good.

**Comment by ximinez** [comment]
> Is `if (value.isString())` inside of `parseUnsigned` a possible code path? It looks like its not but maybe we want to keep it for defensive purposes?

@dangell7 It is, and it's checked by existing test cases. I added more just to be thorough.

File: src/libxrpl/protocol/STParsedJSON.cpp [hunk 1]
```diff
@@ -202,6 +202,173 @@ non_object_in_array(std::string const& item, Json::UInt index)
...
+                        TxFormats::getInstance().findTypeByName(strValue);
+                    ret = detail::make_stvar<STResult>(
+                        field,
+                        Permission::getInstance().txToPermissionType(txType));
+                }
+            }
+            else
+            {
+                ret = detail::make_stvar<STResult>(
+                    field, beast::lexicalCastThrow<Integer>(value.asString()));
```
[the remark this answers is not in the document]
Comment (ximinez) [reply src/libxrpl/protocol/STParsedJSON.cpp:356]: > should this be `field, safe_cast<typename STResult::value_type>(beast::lexicalCastThrow<Integer>( . . .` ?

Since `safe_cast` is essentially a compile-time check, it wouldn't hurt, and might head off some runtime errors.

**Review by Bronek — APPROVED** [review]

**Review by dangell7 — APPROVED** [review]

**Review by yinyiqian1 — APPROVED** [review]
LGTM 👍

**Comment by mvadari** [comment]
> > Should #5726 be merged first so we can ensure that nothing is broken by this PR?
> 
> There are definitely going to be conflicts. We both added the same new file...

Surprisingly, the tests are mostly independent (only overlap is `testArray`). Perhaps some reorg of your tests may be desired, but that's about it.

**Comment by ximinez** [comment]
Commit message
> refactor: Simplify STParsedJSON with some helper functions (#5591)
>
> - Add code coverage for STParsedJSON edge cases
>
> Co-authored-by: Denis Angell <dangell@transia.co>
